go/src/koding/kites/kloud/stack/provider/variable.go

```diff
@@ -67,3 +81,40 @@ func ReplaceVariables(s string, vars []Variable, blank string) string {
 
 	return buf.String()
 }
+
+var escape = func(v *Variable) string {
+	return "$" + v.String()
+}
+
+// EscapeDeadVariables escapes variables which are commented out.
+//
+// The comment format is hardcoded to work for Bash-like shells.
+func EscapeDeadVariables(userdata string) string {
+	var buf bytes.Buffer
+
+	eofNL := strings.HasSuffix(userdata, "\n")
+	scanner := bufio.NewScanner(strings.NewReader(userdata))
+
+	for scanner.Scan() {
+		s := scanner.Text()
+
+		isComment := strings.HasPrefix(strings.TrimSpace(s), "#")
+
+		if !isComment {
+			fmt.Fprintln(&buf, s)
+			continue
+		}
+
+		fmt.Fprintln(&buf, ReplaceVariablesFunc(s, ReadVariables(s), escape))
+	}
+
+	// Scanning from strings.Reader is not going to fail, even if, there's
+	// no recovery from the failure.
+	_ = scanner.Err()
+
+	if eofNL {
+		return buf.String()
+	}
```

Comment:
Not exactly:

If original value ended with a new line - we preserve it. If it doesn't - we trim it, because as you observed - we always write a new line for each of the scanned lines.